newdle/models.py

```diff
@@ -0,0 +1,79 @@
+from pytz import timezone, utc
+from sqlalchemy.dialects.postgresql import JSONB
+from sqlalchemy.schema import CheckConstraint
+
+from newdle.core.db import db
+from newdle.core.util import UTCDateTime, format_dt, parse_dt
+
+
+class Newdle(db.Model):
+    __tablename__ = 'newdles'
+
+    id = db.Column(db.Integer, primary_key=True)
+    creator_uid = db.Column(db.String, nullable=False, index=True)
+    title = db.Column(db.String, nullable=False)
+    duration = db.Column(db.Integer, nullable=False)
+    _timezone = db.Column('timezone', db.String, nullable=False)
+    _time_slots = db.Column('time_slots', JSONB, nullable=False)
+    final_dt = db.Column(UTCDateTime, nullable=True)
+
+    # relationship backrefs:
+    #: - participants (Participant.newdle)
+
+    @property
+    def timezone(self):
+        return timezone(self._timezone)
+
+    @timezone.setter
+    def timezone(self, value):
+        self._timezone = value
+
+    @property
+    def time_slots(self):
+        return [
+            {
+                'start': parse_dt(ts['start']).astimezone(utc),
+                'end': parse_dt(ts['end']).astimezone(utc),
+            }
+            for ts in self._time_slots
+        ]
+
+    @time_slots.setter
+    def time_slots(self, value):
+        self._time_slots = [
+            {
+                'start': format_dt(ts['start'].astimezone(self.timezone)),
+                'end': format_dt(ts['end'].astimezone(self.timezone)),
+            }
+            for ts in value
+        ]
+
+    def __repr__(self):
+        return "<Newdle {} {}>".format(self.id, 'F' if self.final_dt else '')
+
+
+class Participant(db.Model):
+    __tablename__ = 'participants'
+    __table_args__ = (
+        CheckConstraint('email IS NULL OR auth_uid IS NULL', 'email_or_uid_null'),
```

Comment:
don't you want xor logic so one of them has to be set?

Reply:
That was my initial approach, but then I thought about it and the truth table is something like

|---+---+---|
| email | auth_uid |  |
|----|----|----|
|0|0| Ad-hoc user |
|0|1| Directory user |
|1|0| Ad-hoc user with e-mail |
|1|1| Disallowed |

Actually, now that I think of it, should we allow for the third case at all? And how?

Reply:
OK, I've just changed it to `(email IS NULL) = (auth_uid IS NULL)`. I think it makes more sense given the use cases we're targeting right now (either ad-hoc or a directory user, which includes the e-mail).